Obsidian.API/Containers/AbstractContainer.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Obsidian.API
+{
+    public abstract class AbstractContainer : IEnumerable<ItemStack>, IEnumerable
+    {
+        protected ItemStack?[] items;
+
+        public int Size { get; protected set; }
+
+        public InventoryType Type { get; }
+
+        public ChatMessage? Title { get; set; }
+
+        public Guid Uuid { get; private set; } = Guid.NewGuid();
+
+        public List<IPlayer> Viewers { get; private set; } = new List<IPlayer>();
+
+        public ItemStack? this[int index] { get => this.items[index]; set => this.items[index] = value; }
+
+        public AbstractContainer(int size) : this(size, InventoryType.Custom) { }
+
+        internal AbstractContainer(int size, InventoryType type)
+        {
+            this.Size = size;
+            this.Type = type;
+
+            this.items = new ItemStack?[size];
+        }
+
+        //TODO match item meta
+        public virtual int AddItem(ItemStack item)
+        {
+            for (int i = 0; i < this.Size; i++)
+            {
+                var invItem = this.items[i];
+
+                if (invItem?.Type == item.Type)
+                {
+                    if (invItem.Count >= 64)
+                        continue;
+
+                    invItem.Count += item.Count;
+
+                    return i;
+                }
+
+                if (invItem == null)
+                {
+                    this.items[i] = item;
+
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public virtual void SetItem(int slot, ItemStack? item) => this.items[slot] = item;
+
+        public virtual ItemStack? GetItem(int slot) => this.items[slot];
+
+        public virtual bool RemoveItem(int slot, short amount = 1)
```

Comment:
Not sure about this API. I would maybe prefer something like
```cs
public virtual bool RemoveItem(int slot); // Sets slot to null
public virtual bool RemoveItem(int slot, int amount); // Removes specific amount
public virtual bool RemoveItem(int slot, out ItemStack? item); // Sets slot to null and returns the item
public virtual bool RemoveItem(int slot, int amount, out ItemStack? item); // Removes specific amount and returns it
```

Reply:
I originally intended RemoveItem to remove >= 1 of the items amount since there's SetSlot so people can set that slot to null if wanted but I'm fine with something like this as well